### PR TITLE
Modify all user tables to behave in OnDemand manner

### PR DESCRIPTION
**Pull request by dlmarion**
Changes to the on-demand feature that was previously merged into the `elasticity` branch. These changes remove the user option to set a table to an on-demand state and replaces it with:

1. Tablets now have a hosting goal, with values of always, never, and ondemand.
2. The never hosting goal will cause tablets to be unassigned (and unhosted)
3. The always hosting goal will cause tablets to be assigned (and hosted)
4. Users can set the hosting goal on a range of tablets using a new TableOperations method
5. Client requests will insert a column into the tablet metadata to bring ondemand tablets online
6. The OnDemandTabletUnloader will remove the column from the tablet metadata when it unhosts an ondemand tablet.

Note that with these changes certain functions and tests (e.g. BulkIT) do not work. I didn't fix these as we want to move these functions to the Manager anyway.

### CODE REVIEW

File: core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java [hunk 1]
```diff
@@ -1014,39 +1014,15 @@ default TimeType getTimeType(String tableName) throws TableNotFoundException {
...
-   * @return true if table's goal state is online
-   * @since 3.1.0
+   * @param tableName table name
+   * @param range tablet range
+   * @param goal hosting goal (one of {@literal ALWAYS, DEFAULT, NEVER})
+   * @since ELASTICITY_TODO
    */
-  boolean isOnDemand(String tableName) throws AccumuloException, TableNotFoundException;
-
+  default void setTabletHostingGoal(String tableName, Range range, String goal)
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java:1024]: The goal parameter should probably be an enum.  Could use the TabletHostingGoal minus the static methods, maybe have a TabletHostingGoalImpl class that has the static methods.

File: core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java [hunk 2]
```diff
@@ -1014,39 +1014,15 @@ default TimeType getTimeType(String tableName) throws TableNotFoundException {
...
+      throws AccumuloSecurityException, AccumuloException, TableNotFoundException {
+    throw new UnsupportedOperationException();
+  }
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java:1027]: Users will need a way to inspect what hosting goals are currently set.  Could have something like the following.

```java
static class HostingGoalForTablet {
   final TabletId tablet;
   final TabletHostingGoal hostingGoal;
}

/**
 *  Retrieves a stream of tablet hosting goals for tablets that fall within the range.
 */
Stream<HostingGoalForTablet> getTabletHostingGoals(String tableName, Range range)
```

Or could have something that collapses consecutive tablets that have the same goal into the same range.

```java

static class TabletHostingGoalRange {
  final Text startRow;
  final Text endRow;
  final TabletHostingGoal hostingGoal;
}

/**
 *  Retrieves a stream of tablet hosting goals for a tablet.  Consecutive tablets that have same goal are collapsed to form a single TabletHostingGoalRange.
 */
Stream<TabletHostingGoalRange> getTabletHostingGoals(String tableName, Range range)
```

File: core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java [hunk 3]
```diff
@@ -2175,4 +2150,25 @@ private void validatePropertiesToSet(Map<String,String> opts, Map<String,String>
...
+    TableId tableId = context.getTableId(tableName);
+    TabletHostingGoal g = TabletHostingGoal.valueOf(goal.toUpperCase());
+
+    List<TKeyExtent> extents =
+        TabletLocatorImpl.findExtentsForRange(context, tableId, range, Set.of());
+
+    log.debug("Setting tablet hosting goal to {} for extents: {}", goal, extents);
+    ThriftClientTypes.TABLET_MGMT.executeVoid(context,
+        client -> client.setTabletHostingGoal(TraceUtil.traceInfo(), context.rpcCreds(),
+            tableId.canonical(), extents, g.toThrift()));
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java:2171]: Instead of resolving the range into extents on the client side, it would probably be better to pass the range to the manager and let it do the resolution.  Could also make this a FATE operations so that the entire range is processed even in the case of failures.

Making it a FATE operation and passing the range to the manager could be follow on work/PR.

File: core/src/main/java/org/apache/accumulo/core/clientImpl/TabletLocatorImpl.java [hunk 4]
```diff
@@ -542,70 +549,99 @@ public TabletLocation locateTablet(ClientContext context, Text row, boolean skip
...
       String tableName = context.getTableName(tableId);
-      if (!context.tableOperations().isOnDemand(tableName)) {
-        log.trace("bringOnDemandTabletsOnline: table {} is not in ondemand state", tableId);
+      if (!context.tableOperations().isOnline(tableName)) {
+        log.trace("requestTabletHosting: table {} is not online", tableId);
         return;
       }
     } catch (TableNotFoundException e) {
-      log.trace("bringOnDemandTabletsOnline: table not found: {}", tableId);
+      log.trace("requestTabletHosting: table not found: {}", tableId);
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/clientImpl/TabletLocatorImpl.java:587]: This should probably throw an exception.

File: core/src/main/java/org/apache/accumulo/core/clientImpl/TabletLocatorImpl.java [hunk 5]
```diff
@@ -542,70 +549,99 @@ public TabletLocation locateTablet(ClientContext context, Text row, boolean skip
...
       return;
     }
 
-    final Text scanRangeStart = range.getStartKey().getRow();
-    final Text scanRangeEnd = range.getEndKey().getRow();
-    // Turn the scan range into a KeyExtent and bring online all ondemand tablets
+    List<TKeyExtent> extentsToBringOnline =
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/clientImpl/TabletLocatorImpl.java:591]: In #3292 I was trying to minimize the amount of reads to the metadata table.  Like maybe we just read this information to populate the cache and now are reading the metadata table again. I think it would be best to do this its own PR so I don't think any changes are needed in this PR.  I am not completely happy with what I did in #3292 so it would be nice to focus on that one issue rather than do it as part of larger changes.

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/MetadataSchema.java [hunk 6]
```diff
@@ -332,10 +332,16 @@ public static class ExternalCompactionColumnFamily {
...
-    public static class OnDemandAssignmentStateColumnFamily {
-      public static final String STR_NAME = "ondemand";
+    public static class HostingGoalColumnFamily {
+      public static final String STR_NAME = "hosting_goal";
       public static final Text NAME = new Text(STR_NAME);
+      public static final String ALWAYS = "always"; // tablet should always be hosted
+      public static final String DEFAULT = "default"; // tablet should be unassigned
+      public static final String ONDEMAND = "ondemand"; // ondemand request has been made to assign
+                                                        // tablet
+      public static final String NEVER = "never"; // tablet should never be hosted
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/metadata/schema/MetadataSchema.java:342]:
```suggestion
      public static final String ALWAYS = TabletHostingGoal.ALWAYS.name(); // tablet should always be hosted
      public static final String DEFAULT = TabletHostingGoal.DEFAULT.name(); // tablet should be unassigned
      public static final String ONDEMAND = TabletHostingGoal.ONDEMAND.name(); // ondemand request has been made to assign
                                                        // tablet
      public static final String NEVER = TabletHostingGoal.NEVER.name(); // tablet should never be hosted
```

File: core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java [hunk 7]
```diff
@@ -1014,39 +1014,15 @@ default TimeType getTimeType(String tableName) throws TableNotFoundException {
...
-   * <code>onDemand(tableName, true)</code>, this will wait until the table reaches the desired
-   * state before proceeding.
-   *
-   * @param tableName the table to check if in an on-demand state
-   * @throws AccumuloException when there is a general accumulo error
-   * @return true if table's goal state is online
-   * @since 3.1.0
+   * @param tableName table name
+   * @param range tablet range
+   * @param goal hosting goal (one of {@literal ALWAYS, DEFAULT, NEVER})
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java:1021]: Is DEFAULT like ondemand?  I do not have a sense of what it means from the name.  ALWAYS and NEVER are very clear to me.

Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java:1021]: > When a tablet is unloaded, it's hosting_goal column value is set to default.

Seems like this should only change when a user changes it.  If a user specifies they want a tablet to always be hosted and the balancer moves the tablet then its goal state should still be ALWAYS after the move.

> So there are 3 user states and 4 internal states. We need an internal state for on-demand and don't host (default) and on-demand and host (ondemand)

Maybe we could have three user facing states `NEVER,ONDEMAND,ALWAYS` and two columns in the metadata table. A `hosting_goal` column that stores the user facing goal states.  A `hosting_needed` column that is used to indicate a tablet with an ondemand goal state needs hosting.  The `hosting_goal`  is only modified by user operations.  The `hosting_needed` is modified by the system as needed.

> A non-existent hosting_goal column, or one with an empty Value, is treated like default.

 We could also treat any tablet with an unspecified hosting goal as an error if we set them on upgrade and in initialization.  Setting it on upgrade can make the code simpler as it can be written with the expectation that its always set.

Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java:1021]: If we expect the hosting goal to always be set, then modifying the upgrade code to set it could be a follow on PR.

**Comment by keith-turner** [comment]
Users should be able to set the hosting goal in [NewTableConfiguration](https://github.com/apache/accumulo/blob/main/core/src/main/java/org/apache/accumulo/core/client/admin/NewTableConfiguration.java) and have all the tablets in the table start with that goal.  This could be a follow on PR.

File: core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java [hunk 8]
```diff
@@ -2175,4 +2150,25 @@ private void validatePropertiesToSet(Map<String,String> opts, Map<String,String>
...
   }
+
+  @Override
+  public void setTabletHostingGoal(String tableName, Range range, String goal)
+      throws AccumuloSecurityException, AccumuloException, TableNotFoundException {
+    EXISTING_TABLE_NAME.validate(tableName);
+    NOT_BUILTIN_TABLE.validate(tableName);
+    checkArgument(range != null, "range is null");
+    checkArgument(goal != null, "goal is null");
+
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java:2161]: Should this throw an error if the table is offline?  It does not hurt to set it for an offline table, it just has no effect until the table comes online.

**Comment by keith-turner** [comment]
We will need to do something for merging tablets, but not sure what.  Poking around in the code I found this where [tablets are actually merged](https://github.com/apache/accumulo/blob/149b65ce242901ed9a579a8d2b0badc53c24ae04/server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java#L712).  Will need to set the hosting goal for the new tablet based on the hosting goals of the merged tablets. Maybe the merge operation can fail if the hosting goal of the all of the tablets being merged is not the same.  Dealing with merge could be a follow on PR.

File: core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java (hunk 1 above)
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java]: Updated based on your suggestion in d579d7b

File: core/src/main/java/org/apache/accumulo/core/clientImpl/TabletLocatorImpl.java (hunk 4 above)
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/TabletLocatorImpl.java]: Updated based on your suggestion in d579d7b

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/MetadataSchema.java (hunk 6 above)
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/metadata/schema/MetadataSchema.java]: Updated based on your suggestion in d579d7b

File: core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java (hunk 7 above)
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java]: I did the following in d579d7b:

1. Allowed the user to set hosting goal to ALWAYS, NEVER, or ONDEMAND. Removed the DEFAULT option.
2. Stored the user set hosting goal in the tablet metadata in the column `hosting:goal`
3. Default hosting goal value of user tables is ONDEMAND
4. When client requests a Tablet to be hosted, the column `hosting:requested` will be added to the tablet metadata.
5. When the TabletServer unloads an ondemand tablet, it will remove the `hosting:requested` column from the tablet metadata.

File: core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java [hunk 9]
```diff
@@ -1014,39 +1014,15 @@ default TimeType getTimeType(String tableName) throws TableNotFoundException {
...
-   * @return true if table's goal state is online
-   * @since 3.1.0
+   * @param tableName table name
+   * @param range tablet range
+   * @param goal hosting goal
+   * @since ELASTICITY_TODO
    */
-  boolean isOnDemand(String tableName) throws AccumuloException, TableNotFoundException;
-
+  default void setTabletHostingGoal(String tableName, Range range, TabletHostingGoal goal)
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java:1024]: Accumulo APIs that deal with metadata ranges take two Text objects. The range object is a range of row+col. For this operation we only need a row range. If someone sets the columns they would be ignored.  I suppose that since we do not have a RowRange class that is why lots of operations just take two Text objects.

Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java:1024]: Looks like you might have created it already, it's just not a first class citizen.
https://github.com/apache/accumulo/blob/main/core/src/main/java/org/apache/accumulo/core/summary/Gatherer.java#L543

Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java:1024]: I'll move it up and see if I can use it here.

File: core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java [hunk 10]
```diff
@@ -1014,39 +1014,15 @@ default TimeType getTimeType(String tableName) throws TableNotFoundException {
   }
 
   /**
-   * Initiates setting a table to an on-demand state, but does not wait for action to complete
+   * Sets the hosting goal for a range of Tablets in the specified table.
```
[the remark this answers is not in the document]
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/client/admin/TableOperations.java]: Updated in 3410543

File: core/src/main/java/org/apache/accumulo/core/client/admin/TabletHostingGoal.java [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.accumulo.core.client.admin;
+
+public enum TabletHostingGoal {
```
[the remark this answers is not in the document]
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/client/admin/TabletHostingGoal.java:26]: Updated in 3410543

File: server/base/src/main/java/org/apache/accumulo/server/constraints/MetadataConstraints.java [hunk 12]
```diff
@@ -201,7 +203,9 @@ public List<Short> check(Environment env, Mutation mutation) {
       }
 
       if (columnUpdate.getValue().length == 0 && !columnFamily.equals(ScanFileColumnFamily.NAME)
-          && !columnFamily.equals(OnDemandAssignmentStateColumnFamily.NAME)) {
+          && !(columnFamily.equals(HostingColumnFamily.REQUESTED_COLUMN.getColumnFamily())
+              && columnQualifier
+                  .equals(HostingColumnFamily.REQUESTED_COLUMN.getColumnQualifier()))) {
```
Comment (keith-turner) [inline server/base/src/main/java/org/apache/accumulo/server/constraints/MetadataConstraints.java:208]: This can be shortened.

```suggestion
          && !HostingColumnFamily.REQUESTED_COLUMN.equals(columnFamily,columnQualifier)) {
```

File: server/base/src/main/java/org/apache/accumulo/server/constraints/MetadataConstraints.java [hunk 13]
```diff
@@ -215,9 +219,16 @@ public List<Short> check(Environment env, Mutation mutation) {
         } catch (NumberFormatException | ArrayIndexOutOfBoundsException nfe) {
           violations = addViolation(violations, 1);
         }
-      } else if (columnFamily.equals(ScanFileColumnFamily.NAME)) {} else if (columnFamily
-          .equals(OnDemandAssignmentStateColumnFamily.NAME)) {} else if (columnFamily
-              .equals(BulkFileColumnFamily.NAME)) {
+      } else if (columnFamily.equals(ScanFileColumnFamily.NAME)) {
+
+      } else if (columnFamily.equals(HostingColumnFamily.GOAL_COLUMN.getColumnFamily())
+          && columnQualifier.equals(HostingColumnFamily.GOAL_COLUMN.getColumnQualifier())) {
```
Comment (keith-turner) [inline server/base/src/main/java/org/apache/accumulo/server/constraints/MetadataConstraints.java:225]: This can be shortened

```suggestion
      } else if (HostingColumnFamily.GOAL_COLUMN.equals(columnFamily,columnQualifier)) {
```

File: server/base/src/main/java/org/apache/accumulo/server/metadata/TabletMutatorBase.java [hunk 14]
```diff
@@ -269,18 +273,28 @@ protected Mutation getMutation() {
   }
 
   @Override
-  public T putOnDemand() {
-    mutation.put(OnDemandAssignmentStateColumnFamily.STR_NAME, "", "");
+  public T setHostingGoal(TabletHostingGoal goal) {
+    mutation.put(HostingColumnFamily.GOAL_COLUMN.getColumnFamily(),
+        HostingColumnFamily.GOAL_COLUMN.getColumnQualifier(), TabletHostingGoalUtil.toValue(goal));
```
Comment (keith-turner) [inline server/base/src/main/java/org/apache/accumulo/server/metadata/TabletMutatorBase.java:278]:
```suggestion
    HostingColumnFamily.GOAL_COLUMN.put(mutation, TabletHostingGoalUtil.toValue(goal));
```

File: server/base/src/main/java/org/apache/accumulo/server/metadata/TabletMutatorBase.java [hunk 15]
```diff
@@ -269,18 +273,28 @@ protected Mutation getMutation() {
...
+    return getThis();
+  }
+
+  @Override
+  public T setHostingRequested() {
+    mutation.put(HostingColumnFamily.REQUESTED_COLUMN.getColumnFamily(),
+        HostingColumnFamily.REQUESTED_COLUMN.getColumnQualifier(), EMPTY_VALUE);
```
Comment (keith-turner) [inline server/base/src/main/java/org/apache/accumulo/server/metadata/TabletMutatorBase.java:285]:
```suggestion
    HostingColumnFamily.REQUESTED_COLUMN.put(mutation, EMPTY_VALUE);
```

File: server/base/src/main/java/org/apache/accumulo/server/metadata/TabletMutatorBase.java [hunk 16]
```diff
@@ -269,18 +273,28 @@ protected Mutation getMutation() {
...
     return getThis();
   }
 
   @Override
-  public T deleteOnDemand() {
-    mutation.putDelete(OnDemandAssignmentStateColumnFamily.STR_NAME, "");
+  public T deleteHostingRequested() {
+    mutation.putDelete(HostingColumnFamily.REQUESTED_COLUMN.getColumnFamily(),
+        HostingColumnFamily.REQUESTED_COLUMN.getColumnQualifier());
```
Comment (keith-turner) [inline server/base/src/main/java/org/apache/accumulo/server/metadata/TabletMutatorBase.java:292]:
```suggestion
    HostingColumnFamily.REQUESTED_COLUMN.putDelete(mutation);
```

File: server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java [hunk 17]
```diff
@@ -231,7 +232,9 @@ enum CompactionState {
   private final ConcurrentHashMap<Long,List<TabletFile>> bulkImported = new ConcurrentHashMap<>();
 
   private final int logId;
-  private final boolean onDemand;
+
+  // TODO: User can change this, how does it get updated?
```
Comment (keith-turner) [inline server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java:236]: Need to create a follow on issue before merging.

File: server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java [hunk 18]
```diff
@@ -1450,7 +1453,7 @@ public void evaluateOnDemandTabletsForUnload() {
       unloaders.get(tid).evaluate(params);
       onDemandTabletsToUnload.forEach(ke -> {
         log.debug("Unloading on-demand tablet: {} for table: {}", ke, tid);
-        getContext().getAmple().mutateTablet(ke).deleteOnDemand().mutate();
+        getContext().getAmple().mutateTablet(ke).deleteHostingRequested().mutate();
```
Comment (keith-turner) [inline server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java:1456]: Can use one batchwriter for all updates.

```java
      try(var tabletsMutator = getContext().getAmple().mutateTablets()) {
        onDemandTabletsToUnload.forEach(ke -> {
          log.debug("Unloading on-demand tablet: {} for table: {}", ke, tid);
          tabletsMutator.mutateTablet(ke).deleteHostingRequested().mutate();
        });
      }
```

File: test/src/main/java/org/apache/accumulo/test/functional/ManagerAssignmentIT.java [hunk 19]
```diff
@@ -73,26 +77,43 @@ public static void beforeAll() throws Exception {
...
-      // wait for the table to be online
+
+      // wait for the tablet to exist in the metadata table. The tablet
+      // will not be hosted so the current location will be empty.
       TabletLocationState newTablet;
       do {
         UtilWaitThread.sleep(250);
         newTablet = getTabletLocationState(c, tableId);
-      } while (newTablet.current == null);
+      } while (newTablet.extent == null);
```
Comment (keith-turner) [inline test/src/main/java/org/apache/accumulo/test/functional/ManagerAssignmentIT.java:110]: It would be a follow on issue.  These loops could be shortened using the new o.a.a.test.util.Wait class.

File: test/src/main/java/org/apache/accumulo/test/functional/ManagerAssignmentIT.java [hunk 20]
```diff
@@ -102,86 +123,140 @@ public void test() throws Exception {
...
-      assertNull(offline.future);
-      assertNull(offline.current);
-      assertEquals(flushed.getCurrentServer(), offline.getLastServer());
+      // set the hosting goal to always
+      c.tableOperations().setTabletHostingGoal(tableName, new Range(), TabletHostingGoal.ALWAYS);
+      TabletLocationState always;
+      do {
+        UtilWaitThread.sleep(250);
+        always = getTabletLocationState(c, tableId);
+      } while (always.goal == TabletHostingGoal.ONDEMAND);
```
Comment (keith-turner) [inline test/src/main/java/org/apache/accumulo/test/functional/ManagerAssignmentIT.java:153]:
```suggestion
      } while (always.goal != TabletHostingGoal.ALWAYS);
```

File: test/src/main/java/org/apache/accumulo/test/functional/ManagerAssignmentIT.java [hunk 21]
```diff
@@ -102,86 +123,140 @@ public void test() throws Exception {
...
+      assertEquals(flushed.getCurrentServer(), always.getLastServer());
+      assertEquals(TabletHostingGoal.ALWAYS, always.goal);
+
+      // set the hosting goal to never
+      c.tableOperations().setTabletHostingGoal(tableName, new Range(), TabletHostingGoal.NEVER);
+      TabletLocationState never;
+      do {
+        UtilWaitThread.sleep(250);
+        never = getTabletLocationState(c, tableId);
+      } while (never.goal == TabletHostingGoal.ALWAYS);
```
Comment (keith-turner) [inline test/src/main/java/org/apache/accumulo/test/functional/ManagerAssignmentIT.java:166]:
```suggestion
      } while (never.goal != TabletHostingGoal.NEVER);
```

File: test/src/main/java/org/apache/accumulo/test/functional/ManagerAssignmentIT.java [hunk 22]
```diff
@@ -102,86 +123,140 @@ public void test() throws Exception {
...
+      assertEquals(flushed.getCurrentServer(), never.getLastServer());
+      assertEquals(TabletHostingGoal.NEVER, never.goal);
+
+      // set the hosting goal to ondemand
+      c.tableOperations().setTabletHostingGoal(tableName, new Range(), TabletHostingGoal.ONDEMAND);
+      TabletLocationState ondemand;
+      do {
+        UtilWaitThread.sleep(250);
+        ondemand = getTabletLocationState(c, tableId);
+      } while (ondemand.goal == TabletHostingGoal.NEVER);
```
Comment (keith-turner) [inline test/src/main/java/org/apache/accumulo/test/functional/ManagerAssignmentIT.java:179]:
```suggestion
      } while (ondemand.goal != TabletHostingGoal.ONDEMAND);
```

File: core/src/main/java/org/apache/accumulo/core/clientImpl/TabletLocatorImpl.java [hunk 23]
```diff
@@ -542,70 +556,95 @@ public TabletLocation locateTablet(ClientContext context, Text row, boolean skip
...
         // the end row of this tablet is before the start row, skip it
         log.trace("tablet {} is before scan start range: {}", tabletExtent, scanRangeStart);
         continue;
       }
-      if (tm.getPrevEndRow() != null && tm.getPrevEndRow().compareTo(scanRangeEnd) > 0) {
+      if (scanRangeEnd != null && tm.getPrevEndRow() != null
+          && tm.getPrevEndRow().compareTo(scanRangeEnd) > 0) {
         // the start row of this tablet is after the scan range end row, skip it
         log.trace("tablet {} is after scan end range: {}", tabletExtent, scanRangeEnd);
         continue;
```
[the remark this answers is not in the document]
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/TabletLocatorImpl.java:634]: If ample returns the tablet metadata in order, then that probably makes sense.

File: test/src/main/java/org/apache/accumulo/test/functional/ManagerAssignmentIT.java (hunk 22 above)
Comment (dlmarion) [reply test/src/main/java/org/apache/accumulo/test/functional/ManagerAssignmentIT.java]: Suggestion applied in cd223b6

File: test/src/main/java/org/apache/accumulo/test/functional/ManagerAssignmentIT.java (hunk 21 above)
Comment (dlmarion) [reply test/src/main/java/org/apache/accumulo/test/functional/ManagerAssignmentIT.java]: Suggestion applied in cd223b6

File: test/src/main/java/org/apache/accumulo/test/functional/ManagerAssignmentIT.java (hunk 20 above)
Comment (dlmarion) [reply test/src/main/java/org/apache/accumulo/test/functional/ManagerAssignmentIT.java]: Suggestion applied in cd223b6

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/MetadataSchema.java [hunk 24]
```diff
@@ -357,10 +358,18 @@ public static class ExternalCompactionColumnFamily {
...
-    public static class OnDemandAssignmentStateColumnFamily {
-      public static final String STR_NAME = "ondemand";
+    public static class HostingColumnFamily {
+      public static final String STR_NAME = "hosting";
       public static final Text NAME = new Text(STR_NAME);
+      public static final String GOAL_QUAL = "goal";
+      public static final ColumnFQ GOAL_COLUMN = new ColumnFQ(NAME, new Text(GOAL_QUAL));
+      public static final String ALWAYS_GOAL = TabletHostingGoal.ALWAYS.name();
+      public static final String ONDEMAND_GOAL = TabletHostingGoal.ONDEMAND.name();
+      public static final String NEVER_GOAL = TabletHostingGoal.NEVER.name();
```
[the remark this answers is not in the document]
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/metadata/schema/MetadataSchema.java]: Suggestion applied in cd223b6

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletsMetadata.java [hunk 25]
```diff
@@ -283,6 +283,9 @@ public Options fetch(ColumnType... colsToFetch) {
           case FLUSH_ID:
             qualifiers.add(FLUSH_COLUMN);
             break;
+          case HOSTING_GOAL:
+            families.add(HostingColumnFamily.NAME);
+            break;
```
[the remark this answers is not in the document]
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletsMetadata.java]: Suggestion applied in cd223b6

File: server/base/src/main/java/org/apache/accumulo/server/constraints/MetadataConstraints.java [hunk 26]
```diff
@@ -100,7 +101,7 @@ public class MetadataConstraints implements Constraint {
           ChoppedColumnFamily.NAME,
           ClonedColumnFamily.NAME,
           ExternalCompactionColumnFamily.NAME,
-          OnDemandAssignmentStateColumnFamily.NAME);
+          HostingColumnFamily.NAME);
```
[the remark this answers is not in the document]
Comment (dlmarion) [reply server/base/src/main/java/org/apache/accumulo/server/constraints/MetadataConstraints.java]: Suggestion applied in cd223b6

File: server/base/src/main/java/org/apache/accumulo/server/constraints/MetadataConstraints.java (hunk 12 above)
Comment (dlmarion) [reply server/base/src/main/java/org/apache/accumulo/server/constraints/MetadataConstraints.java]: Suggestion applied in cd223b6

File: server/base/src/main/java/org/apache/accumulo/server/constraints/MetadataConstraints.java (hunk 13 above)
Comment (dlmarion) [reply server/base/src/main/java/org/apache/accumulo/server/constraints/MetadataConstraints.java]: Suggestion applied in cd223b6

File: server/base/src/main/java/org/apache/accumulo/server/metadata/TabletMutatorBase.java (hunk 14 above)
Comment (dlmarion) [reply server/base/src/main/java/org/apache/accumulo/server/metadata/TabletMutatorBase.java]: Suggestion applied in cd223b6

File: server/base/src/main/java/org/apache/accumulo/server/metadata/TabletMutatorBase.java (hunk 15 above)
Comment (dlmarion) [reply server/base/src/main/java/org/apache/accumulo/server/metadata/TabletMutatorBase.java]: Suggestion applied in cd223b6

File: core/src/main/java/org/apache/accumulo/core/clientImpl/TabletLocatorImpl.java (hunk 23 above)
[the remark this answers is not in the document]
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/TabletLocatorImpl.java]: @keith-turner  - I attempted to fix the logic in cd223b6 to account for the exclusive end row. Please take a look when you get a chance.

File: core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java [hunk 27]
```diff
@@ -1379,8 +1383,10 @@ private void waitForTableStateTransition(TableId tableId, TableState expectedSta
       for (TabletMetadata tablet : tablets) {
         total++;
         Location loc = tablet.getLocation();
+        TabletHostingGoal goal = tablet.getHostingGoal();
 
         if ((expectedState == TableState.ONLINE
+            && (goal == TabletHostingGoal.ALWAYS || goal == TabletHostingGoal.ONDEMAND)
```
[the remark this answers is not in the document]
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java]: First suggestion applied in cd223b6

File: core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java [hunk 28]
```diff
@@ -1399,7 +1405,8 @@ private void waitForTableStateTransition(TableId tableId, TableState expectedSta
               "Saw unexpected table Id " + tableId + " " + tablet.getExtent());
         }
 
-        if (lastExtent != null && !tablet.getExtent().isPreviousExtent(lastExtent)) {
+        if (expectedState == TableState.OFFLINE && lastExtent != null
```
[the remark this answers is not in the document]
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java]: Removed the offline check in cd223b6

File: server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java (hunk 17 above)
Comment (dlmarion) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java:236]: captured this in the comment of follow-on tasks

File: test/src/main/java/org/apache/accumulo/test/functional/ManagerAssignmentIT.java (hunk 19 above)
Comment (dlmarion) [reply test/src/main/java/org/apache/accumulo/test/functional/ManagerAssignmentIT.java:110]: Captured in comment of follow-on tasks

File: core/src/main/java/org/apache/accumulo/core/clientImpl/TabletLocatorImpl.java (hunk 23 above)
[the remark this answers is not in the document]
Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/TabletLocatorImpl.java]: @dlmarion  take a look at https://github.com/keith-turner/accumulo/commit/1255dfbaf2cfcd1c4de4409d3214e2586120eee0 where I made an attempt at this.  

I was looking at this code and the end row was giving me a headache because the inclusivity  of the key does not easily map to the inclusivity of the end row for different cases.  I realized that the batch scanner has to map ranges to tablets and [looked at what it did](https://github.com/apache/accumulo/blob/15c9f55280f4ce28a80b800f1e05128f33c85db3/core/src/main/java/org/apache/accumulo/core/clientImpl/TabletLocatorImpl.java#L327).  That code uses range.afterEndKey so I gave that a try. The code for the batch scanner is slightly different (its trying to determine if it should advance to the next tablet, where this loop is concerned with the current tablet), but I think the range.afterEndKey function could still be used in this code.

I still don't feel 100% about the code in my commit.  Range overlapping code is tricky and needs tests for all of the edge cases.  If this commit works, maybe its ok for now w/o more test because I would like to better integrate the code that hosts tablets ondemand w/ the cache reads of the metadata (and maybe part of that could be leverage the batch scanner cache code that does have lots of unit test).  We may need a follow on issue for test in the case where that work is not done.

**Review by keith-turner — APPROVED** [review]
One more follow on to add the list for this PR would be ITs that rigorously test the boundry conditions for tablet with different goal states.  So for the batch scanner, batch writer, and scanner try interacting with rows right around the tablet boundries for tablets with different goal states.  Like the first row and last row in a tablet with a different goal than the tablet next to it.
